File: ascii_fields/animations/orbitals.py
import math

from ..core import Animation, clamp, render_field


def _1s(x, z, r):
  return math.exp(-r)


def _2s(x, z, r):
  return (1.0 - 0.5 * r) * math.exp(-0.5 * r)


def _2pz(x, z, r):
  return z * math.exp(-0.5 * r)


def _3dz2(x, z, r):
  return (3.0 * z * z - r * r) * math.exp(-r / 3.0)


def _3dxz(x, z, r):
  return x * z * math.exp(-r / 3.0)


def _4fz3(x, z, r):
  return z * (5.0 * z * z - 3.0 * r * r) * math.exp(-0.25 * r)


# (label, amplitude fn, world half-extent in Bohr radii)
ORBITALS = [
  ("1s", _1s, 6.0),
  ("2p", _2pz, 12.0),
  ("2s", _2s, 14.0),
  ("3d_z2", _3dz2, 20.0),
  ("3d_xz", _3dxz, 20.0),
  ("4f", _4fz3, 26.0),
]

STATE_SECONDS = 4.5
TRANSITION = 1.6


class OrbitalsAnimation(Animation):
  """Hydrogen electron probability clouds |psi|^2 in a cross-section, slowly
  morphing through real (n, l, m) orbitals: 1s -> 2p -> 2s -> 3d -> 4f ..."""

  thresholds = [
    (0.05, " "), (0.14, "."), (0.26, ":"), (0.40, "-"), (0.54, "="),
    (0.68, "+"), (0.80, "*"), (0.90, "#"), (1.01, "@"),
  ]

  def __init__(self):
    self._norm = {}

  def _normalizer(self, index):
    if index not in self._norm:
      _, fn, extent = ORBITALS[index]
      peak = 1e-9
      for i in range(41):
        for j in range(41):
          x = (i / 40.0 - 0.5) * 2.0 * extent
          z = (j / 40.0 - 0.5) * 2.0 * extent
          a = fn(x, z, math.hypot(x, z))
          peak = max(peak, a * a)
      self._norm[index] = 1.0 / peak
    return self._norm[index]

  def render(self, width, height, elapsed, phase, options):
    cycle = STATE_SECONDS
    pos = elapsed / cycle
    i0 = int(pos) % len(ORBITALS)
    i1 = (i0 + 1) % len(ORBITALS)
    frac = pos - math.floor(pos)
    blend = clamp((frac - (1.0 - TRANSITION / cycle)) / (TRANSITION / cycle))

    label0, fn0, ext0 = ORBITALS[i0]
    label1, fn1, ext1 = ORBITALS[i1]
    n0 = self._normalizer(i0)
    n1 = self._normalizer(i1)
    contrast = options.contrast
    grid = []
    for row in range(height):
      pz = (row - (height - 1) * 0.5) / max(1.0, height * 0.5)
      line = []
      for col in range(width):
        px = (col - (width - 1) * 0.5) / max(1.0, width * 0.5)
        x0, z0 = px * ext0, pz * ext0
        a0 = fn0(x0, z0, math.hypot(x0, z0))
        d0 = a0 * a0 * n0
        if blend > 0.0:
          x1, z1 = px * ext1, pz * ext1
          a1 = fn1(x1, z1, math.hypot(x1, z1))
          d1 = a1 * a1 * n1
          d = d0 * (1.0 - blend) + d1 * blend
        else:
          d = d0
        level = (clamp(d)) ** 0.45
        line.append(clamp(level * contrast))
      grid.append(line)
    return render_field(width, height, grid, options, self)

File: src/animations/orbitals.rs
use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_field, FieldStyle};
use std::collections::HashMap;

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "nebula" };
const TH: &[(f64, char)] = &[
  (0.05, ' '),
  (0.14, '.'),
  (0.26, ':'),
  (0.40, '-'),
  (0.54, '='),
  (0.68, '+'),
  (0.80, '*'),
  (0.90, '#'),
  (1.01, '@'),
];

type OrbFn = fn(f64, f64, f64) -> f64;
// (label, function, world half-extent)
const ORBITALS: &[(&str, OrbFn, f64)] = &[
  ("1s", orb_1s, 6.0),
  ("2p", orb_2pz, 12.0),
  ("2s", orb_2s, 14.0),
  ("3d_z2", orb_3dz2, 20.0),
  ("3d_xz", orb_3dxz, 20.0),
  ("4f", orb_4fz3, 26.0),
];
const STATE_SECONDS: f64 = 4.5;
const TRANSITION: f64 = 1.6;

fn orb_1s(_x: f64, _z: f64, r: f64) -> f64 {
  (-r).exp()
}
fn orb_2s(_x: f64, _z: f64, r: f64) -> f64 {
  (1.0 - 0.5 * r) * (-0.5 * r).exp()
}
fn orb_2pz(_x: f64, z: f64, r: f64) -> f64 {
  z * (-0.5 * r).exp()
}
fn orb_3dz2(_x: f64, z: f64, r: f64) -> f64 {
  (3.0 * z * z - r * r) * (-r / 3.0).exp()
}
fn orb_3dxz(x: f64, z: f64, r: f64) -> f64 {
  x * z * (-r / 3.0).exp()
}
fn orb_4fz3(_x: f64, z: f64, r: f64) -> f64 {
  z * (5.0 * z * z - 3.0 * r * r) * (-0.25 * r).exp()
}

pub struct Orbitals {
  norm: HashMap<usize, f64>,
}
impl Default for Orbitals {
  fn default() -> Self {
    Self { norm: HashMap::new() }
  }
}

impl Orbitals {
  fn normalizer(&mut self, idx: usize) -> f64 {
    if let Some(&v) = self.norm.get(&idx) {
      return v;
    }
    let (_, f, ext) = ORBITALS[idx];
    let mut peak: f64 = 1e-9;
    for i in 0..=40 {
      for j in 0..=40 {
        let x = (i as f64 / 40.0 - 0.5) * 2.0 * ext;
        let z = (j as f64 / 40.0 - 0.5) * 2.0 * ext;
        let a = f(x, z, (x * x + z * z).sqrt());
        if a * a > peak {
          peak = a * a;
        }
      }
    }
    let n = 1.0 / peak;
    self.norm.insert(idx, n);
    n
  }
}

impl Animation for Orbitals {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    let w = ctx.width;
    let h = ctx.height;
    let cycle = STATE_SECONDS;
    let pos = ctx.elapsed / cycle;
    let i0 = (pos as usize) % ORBITALS.len();
    let i1 = (i0 + 1) % ORBITALS.len();
    let frac = pos - pos.floor();
    let blend = clamp((frac - (1.0 - TRANSITION / cycle)) / (TRANSITION / cycle));
    let (_, f0, ext0) = ORBITALS[i0];
    let (_, f1, ext1) = ORBITALS[i1];
    let n0 = self.normalizer(i0);
    let n1 = self.normalizer(i1);
    let contrast = ctx.options.contrast;
    let mut grid = vec![0.0_f64; w * h];
    for row in 0..h {
      let pz = (row as f64 - (h as f64 - 1.0) * 0.5) / (h as f64 * 0.5).max(1.0);
      let base = row * w;
      for col in 0..w {
        let px = (col as f64 - (w as f64 - 1.0) * 0.5) / (w as f64 * 0.5).max(1.0);
        let x0 = px * ext0;
        let z0 = pz * ext0;
        let a0 = f0(x0, z0, (x0 * x0 + z0 * z0).sqrt());
        let d0 = a0 * a0 * n0;
        let d = if blend > 0.0 {
          let x1 = px * ext1;
          let z1 = pz * ext1;
          let a1 = f1(x1, z1, (x1 * x1 + z1 * z1).sqrt());
          d0 * (1.0 - blend) + a1 * a1 * n1 * blend
        } else {
          d0
        };
        let level = clamp(d).powf(0.45);
        grid[base + col] = clamp(level * contrast);
      }
    }
    render_field(ctx, &grid, TH, &STYLE, out);
  }
}
